mp4parse/src/lib.rs

```diff
@@ -1302,45 +1445,71 @@ pub fn read_avif<T: Read>(f: &mut T) -> Result<AvifContext> {
         })
     });
 
-    let mut primary_item = TryVec::new();
+    let mut primary_item = None;
     let mut alpha_item = None;
 
-    // load data of relevant items
-    for loc in meta.iloc_items.iter() {
-        let mut item_data = if loc.item_id == meta.primary_item_id {
+    // store data or record location of relevant items
+    for loc in meta.iloc_items {
+        let item = if loc.item_id == meta.primary_item_id {
             &mut primary_item
         } else if Some(loc.item_id) == alpha_item_id {
-            alpha_item.get_or_insert_with(TryVec::new)
+            &mut alpha_item
         } else {
             continue;
         };
 
         if loc.construction_method != ConstructionMethod::File {
             return Err(Error::Unsupported("unsupported construction_method"));
         }
-        for extent in loc.extents.iter() {
+
+        assert!(item.is_none());
+
+        // If our item is spread over multiple extents, we'll need to copy it
+        // into a contiguous buffer. Otherwise, we can just store the extent
+        // and return a pointer into the mdat later to avoid the copy.
+        if loc.extents.len() > 1 {
+            *item = Some(AvifItem::Data(TryVec::new()));
+        }
+
+        for extent in loc.extents {
             let mut found = false;
-            // try to find an overlapping mdat
-            for mdat in mdats.iter_mut() {
-                if mdat.matches_extent(&extent.extent_range) {
-                    item_data.append(&mut mdat.data)?;
-                    found = true;
-                    break;
-                } else if mdat.contains_extent(&extent.extent_range) {
-                    mdat.read_extent(&extent.extent_range, &mut item_data)?;
+            // try to find an mdat which contains the extent
+            for mdat in item_storage.iter_mut() {
+                if let Some(extent_slice) = mdat.get(&extent) {
+                    match item {
+                        None => {
+                            trace!("Using AvifItem::Location");
+                            *item = Some(AvifItem::Location(extent));
+                        }
+                        Some(AvifItem::Data(item_data)) => {
+                            trace!("Using AvifItem::Data");
+                            item_data.extend_from_slice(extent_slice)?;
+                        }
+                        _ => unreachable!(),
+                    }
                     found = true;
                     break;
                 }
             }
+
             if !found {
                 return Err(Error::InvalidData(
-                    "iloc contains an extent that is not in mdat",
+                    "iloc contains an extent that is not in any mdat",
                 ));
             }
         }
```

Comment:
I considered this alternative implementation. I'm curious which approach others find most readable:
```rust
        if loc.extents.len() == 1 {
            let extent = loc.extents.pop().expect("exactly one extent");
            for mdat in &item_storage {
                if mdat.get(&extent).is_some() {
                    *item = Some(AvifItem::Location(extent));
                    break;
                }
            }

            if item.is_none() {
                return Err(Error::InvalidData(
                    "iloc contains an extent that is not in any mdat",
                ));
            }
        } else {
            let mut found = false;
            let mut item_data = TryVec::new();

            for extent in loc.extents {
                // try to find an mdat which contains the extent
                for mdat in item_storage.iter_mut() {
                    if let Some(extent_slice) = mdat.get(&extent) {
                        item_data.extend_from_slice(extent_slice)?;
                        found = true;
                        break;
                    }
                }

                if !found {
                    return Err(Error::InvalidData(
                        "iloc contains an extent that is not in any mdat",
                    ));
                }
            }

            *item = Some(AvifItem::Data(item_data));
        }
```

Reply:
The version in the commit seems nice, but no strong preference.